### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by jmaitrehenry** [containous/traefik#2765]
### Do you want to request a *feature* or report a *bug*?
Bug

### What did you do?
Update Traefik to 1.5 with HTTP challenge.

### What did you expect to see?

No panic

### What did you see instead?

```
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    | panic: runtime error: invalid memory address or nil pointer dereference
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    | [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x9e9417]
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    |
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    | goroutine 123 [running]:
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    | github.com/containous/traefik/vendor/github.com/xenolf/lego/acme.(*Client).getChallenges.func1(0x0, 0xc42083a600, 0xc42083a5a0, 0xc4205468b0, 0xb)
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    | 	/go/src/github.com/containous/traefik/vendor/github.com/xenolf/lego/acme/client.go:550 +0xd7
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    | created by github.com/containous/traefik/vendor/github.com/xenolf/lego/acme.(*Client).getChallenges
traefik_traefik.0.msw1d7gplg0e@vm-mon-docker22    | 	/go/src/github.com/containous/traefik/vendor/github.com/xenolf/lego/acme/client.go:547 +0x133
traefik_traefik.0.uix0fa50u95a@vm-mon-docker20    | time="2018-01-25T15:31:27Z" level=error msg="Datastore sync error: Object lock value: expected f1ec4385-0616-4d5a-83c1-6796896c8b12, got , retrying in 521.065087ms"
traefik_traefik.0.kfc64oj0v8h4@vm-mon-docker21    | time="2018-01-25T15:31:27Z" level=error msg="Cannot unmarshall private key []"
traefik_traefik.0.kfc64oj0v8h4@vm-mon-docker21    | time="2018-01-25T15:31:27Z" level=error msg="Error building ACME client &{Email: Registration:<nil> PrivateKey:[] DomainsCertificate:{Certs:[] lock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0}} ChallengeCerts:map[] HTTPChallenge:map[]}: private key was nil"
```

### Output of `traefik version`: (_What version of Traefik are you using?_)
1.5

### What is your environment & configuration (arguments, toml, provider, platform, ...)?
```
      - "--logLevel=DEBUG"
      - "--web"
      - "--entrypoints=Name:http Address::80 Redirect.EntryPoint:https"
      - "--entrypoints=Name:https Address::443 TLS"
      - "--defaultentrypoints=http,https"
      - "--acme"
      - "--acme.storage=traefik/acme/account"
      - "--acme.entryPoint=https"
      - "--acme.OnHostRule=true"
      - "--acme.onDemand=true"
      - "--acme.email=julien.maitrehenry@me.com"
      - "--acme.httpChallenge"
      - "--acme.httpChallenge.entryPoint=http"
      - "--docker"
      - "--docker.swarmmode"
      - "--docker.domain=jmaitrehenry.ca"
      - "--docker.watch"
      - "--consul"
      - "--consul.endpoint=consul:8500"
      - "--consul.prefix=traefik"
      - "--debug"
```

**Comment by nmengin** [comment]
Hello @jmaitrehenry.

Can you give us more information about your configuration?

- Did you use the configuration given in your issue when you started Træfik or when you executed the `storeconfig` command?
- Is your configuration stored in your Consul? Is it a new configuration or are you using the same configuration as in previous version? Did you update it with the new options?
- Can you provide us all the logs?

Indeed, your ACME account seems to be partially initialized ans it's maybe the origin of the `panic`.
In function of your configuration we can analyze the problem and, I hope, give you quickly a solution.

**Comment by gytisgreitai** [comment]
I'm probably having the same issue.

```

time="2018-02-15T10:47:07Z" level=debug msg="Looking for provided certificate to validate [traefik.xxxx.eu]..."
time="2018-02-15T10:47:07Z" level=debug msg="No provided certificate found for domains [traefik.xxxx.eu], get ACME certificate."
time="2018-02-15T10:47:07Z" level=debug msg="Looking for an existing ACME challenge for traefik.xxxx.eu..."
time="2018-02-15T10:47:07Z" level=debug msg="Loading ACME certificates [traefik.xxxx.eu]..."
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x9e9d87]

goroutine 27 [running]:
github.com/containous/traefik/vendor/github.com/xenolf/lego/acme.(*Client).getChallenges.func1(0x0, 0xc420b1e840, 0xc420b1e7e0, 0xc4203683a0, 0x12)
	/go/src/github.com/containous/traefik/vendor/github.com/xenolf/lego/acme/client.go:550 +0xd7
created by github.com/containous/traefik/vendor/github.com/xenolf/lego/acme.(*Client).getChallenges
	/go/src/github.com/containous/traefik/vendor/github.com/xenolf/lego/acme/client.go:547 +0x133
```

I'm running on `kubernetes` with `zookeeper` backend. My deployment has an init container with the following configuration which should store the config into zookeeper:
```

        - storeconfig
        - --api
        - --debug
        - --defaultentrypoints=http,https
        - --entrypoints=Name:http Address::80
        - --entrypoints=Name:https Address::443 TLS:/etc/nginx/yyyy-io-ssl/nginx.crt,/etc/nginx/yyyy-io-ssl/nginx.key
        - --zookeeper
        - --zookeeper.endpoint=zookeeper.test.svc.cluster.local:2181
        - --acme
        - --acme.email=hostmaster@yyyy.io
        - --acme.storage=traefik/acme/account
        - --acme.entrypoint=https
        - --acme.OnDemand=true
        - --acme.httpchallenge
        - --acme.onHostRule=true
        - --acme.httpchallenge.entryPoint=http
        - --acme.caServer=https://acme-staging.api.letsencrypt.org/directory
```

And then the actual traefik configuration for the Deployment:

```
        - --debug
        - --zookeeper
        - --zookeeper.endpoint=zookeeper.test.svc.cluster.local:2181
```

ZK contents (pushed via zookeeper client), retrieved via `/api/provider/zk`:

```
{  
   "zk":{  
      "backends":{  
         "front-service":{  
            "servers":{  
               "front-service":{  
                  "url":"http://front-service-service:80",
                  "weight":1
               }
            },
            "loadBalancer":{  
               "method":"wrr"
            }
         }
      },
      "frontends":{  
         "traefik.xxxx.eu":{  
            "entryPoints":[  
               "http",
               "https"
            ],
            "backend":"front-service",
            "routes":{  
               "traefik.xxxx.eu":{  
                  "rule":"Host:traefik.xxxx.eu"
               }
            },
            "passHostHeader":true,
            "priority":0,
            "basicAuth":null
         },
         "traefik.yyyy.io":{  
            "entryPoints":[  
               "http",
               "https"
            ],
            "backend":"front-service",
            "routes":{  
               "traefik.yyyy.io":{  
                  "rule":"Host:traefik.yyyy.io"
               }
            },
            "passHostHeader":true,
            "priority":0,
            "basicAuth":null
         }
      }
   }
}


```

Full config from debug log:

```
{
   "LifeCycle":{
      "RequestAcceptGraceTimeout":0,
      "GraceTimeOut":0
   },
   "GraceTimeOut":0,
   "Debug":true,
   "CheckNewVersion":true,
   "SendAnonymousUsage":false,
   "AccessLogsFile":"",
   "AccessLog":null,
   "TraefikLogsFile":"",
   "TraefikLog":null,
   "LogLevel":"DEBUG",
   "EntryPoints":{
      "http":{
         "Network":"",
         "Address":":80",
         "TLS":null,
         "Redirect":null,
         "Auth":null,
         "WhitelistSourceRange":null,
         "Compress":false,
         "ProxyProtocol":null,
         "ForwardedHeaders":{
            "Insecure":true,
            "TrustedIPs":null
         }
      },
      "https":{
         "Network":"",
         "Address":":443",
         "TLS":{
            "MinVersion":"",
            "CipherSuites":null,
            "Certificates":[
               {
                  "CertFile":"/etc/nginx/yyyy-io-ssl/nginx.crt",
                  "KeyFile":"/etc/nginx/yyyy-io-ssl/nginx.key"
               }
            ],
            "ClientCAFiles":null,
            "ClientCA":{
               "Files":null,
               "Optional":false
            }
         },
         "Redirect":null,
         "Auth":null,
         "WhitelistSourceRange":null,
         "Compress":false,
         "ProxyProtocol":null,
         "ForwardedHeaders":{
            "Insecure":true,
            "TrustedIPs":null
         }
      },
      "traefik":{
         "Network":"",
         "Address":":8080",
         "TLS":null,
         "Redirect":null,
         "Auth":null,
         "WhitelistSourceRange":null,
         "Compress":false,
         "ProxyProtocol":null,
         "ForwardedHeaders":{
            "Insecure":true,
            "TrustedIPs":null
         }
      }
   },
   "Cluster":{
      "Node":"c62a87cc-fd60-4e99-929f-352ac1961ce3",
      "Store":{
         "Store":{

         },
         "Prefix":"traefik"
      }
   },
   "Constraints":[

   ],
   "ACME":{
      "Email":"hostmaster@yyyy.io",
      "Domains":null,
      "Storage":"traefik/acme/account",
      "StorageFile":"",
      "OnDemand":true,
      "OnHostRule":true,
      "CAServer":"https://acme-staging.api.letsencrypt.org/directory",
      "EntryPoint":"https",
      "DNSChallenge":null,
      "HTTPChallenge":{
         "EntryPoint":"http"
      },
      "DNSProvider":"",
      "DelayDontCheckDNS":0,
      "ACMELogging":false,
      "TLSConfig":null
   },
   "DefaultEntryPoints":[
      "http",
      "https"
   ],
   "ProvidersThrottleDuration":2000000000,
   "MaxIdleConnsPerHost":200,
   "IdleTimeout":0,
   "InsecureSkipVerify":false,
   "RootCAs":null,
   "Retry":null,
   "HealthCheck":{
      "Interval":30000000000
   },
   "RespondingTimeouts":null,
   "ForwardingTimeouts":null,
   "Web":null,
   "Docker":null,
   "File":null,
   "Marathon":null,
   "Consul":null,
   "ConsulCatalog":null,
   "Etcd":null,
   "Zookeeper":{
      "Watch":true,
      "Filename":"",
      "Constraints":[

      ],
      "Trace":false,
      "DebugLogGeneratedTemplate":false,
      "Endpoint":"zookeeper.test.svc.cluster.local:2181",
      "Prefix":"traefik",
      "TLS":null,
      "Username":"",
      "Password":""
   },
   "Boltdb":null,
   "Kubernetes":null,
   "Mesos":null,
   "Eureka":null,
   "ECS":null,
   "Rancher":null,
   "DynamoDB":null,
   "ServiceFabric":null,
   "Rest":null,
   "API":{
      "EntryPoint":"traefik",
      "Dashboard":true,
      "Debug":true,
      "CurrentConfigurations":null,
      "Statistics":null
   },
   "Metrics":null,
   "Ping":null
}

```
I've tested with two domains: yyyy.io for which I have a wildcard and don't need ACME, and xxxx.eu for which I need acme.
The problem only happens with `onDemand=true`. Though if that deprecated flag is turned off, no certificates are issued for `traefik.xxxx.eu` domain:

```
time="2018-02-15T11:06:02Z" level=debug msg="Looking for provided certificate to validate [traefik.xxxx.eu]..."
time="2018-02-15T11:06:02Z" level=debug msg="No provided certificate found for domains [traefik.xxxx.eu], get ACME certificate."
time="2018-02-15T11:06:02Z" level=debug msg="Looking for an existing ACME challenge for traefik.xxxx.eu..."
time="2018-02-15T11:06:02Z" level=debug msg="No certificate found or generated for traefik.xxxx.eu"


```

**Comment by brodkin** [comment]
+1 Same issue here. Previously used ACME HTTP challenge with a file and attempted to switch to init container with Consul. Using the provided documentation, the containers start with a self-signed certificate, which is never replaced across multiple restarts of the init or Traefik containers.

Attempting to work around this, switching on the depreciated `onDemand` results in the panic.

**Comment by brodkin** [comment]
P.S. Switching back to the file method for ACME still works so it's not something in the environment that has changed to cause this.

**Comment by nmengin** [comment]
Hello @jmaitrehenry , @gytisgreitai , @brodkin 

Did you reproduce the issue with new Træfik version?
Indeed, we upgraded the LEGO version and that might fix the problem.

**Comment by gytisgreitai** [comment]
sorry, I've moved on with nginx + lua-resty-auto-ssl

**Comment by traefiker** [comment]
Closed by #3888.
